File: backend/app/schemas/medicine.py
import uuid
from datetime import datetime, time

from pydantic import BaseModel, Field


class MedicineCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    dosage: str = Field(min_length=1, max_length=255)
    scheduled_time: time
    frequency: str = Field(default="daily", max_length=32)
    active: bool = True
    reminder_enabled: bool = True
    pill_count: int | None = Field(default=None, ge=0)


class MedicineUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    dosage: str | None = Field(default=None, min_length=1, max_length=255)
    scheduled_time: time | None = None
    frequency: str | None = Field(default=None, max_length=32)
    active: bool | None = None
    reminder_enabled: bool | None = None
    pill_count: int | None = Field(default=None, ge=0)


class LabelPreviewOut(BaseModel):
    product_name: str | None = None
    strength: str | None = None
    form: str | None = None
    summary: str | None = None


class MedicineOut(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    dosage: str
    scheduled_time: time
    frequency: str
    active: bool
    reminder_enabled: bool
    pill_count: int | None
    label_image_key: str | None = None
    label_analysis_text: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/pillpal_app/lib/models/medicine.dart
import 'package:flutter/material.dart';

class Medicine {
  final String id;
  final String userId;
  final String name;
  final String dosage;
  final TimeOfDay scheduledTime;
  final String frequency;
  final bool active;
  final bool reminderEnabled;
  final int? pillCount;
  final String? labelImageKey;
  final String? labelAnalysisText;
  final DateTime createdAt;

  Medicine({
    required this.id,
    required this.userId,
    required this.name,
    required this.dosage,
    required this.scheduledTime,
    required this.frequency,
    required this.active,
    this.reminderEnabled = true,
    this.pillCount,
    this.labelImageKey,
    this.labelAnalysisText,
    required this.createdAt,
  });

  factory Medicine.fromJson(Map<String, dynamic> json) {
    final timeParts = json['scheduled_time'].toString().split(':');
    return Medicine(
      id: json['id'].toString(),
      userId: json['user_id'].toString(),
      name: json['name'] as String,
      dosage: json['dosage'] as String,
      scheduledTime: TimeOfDay(
        hour: int.parse(timeParts[0]),
        minute: int.parse(timeParts[1]),
      ),
      frequency: json['frequency'] as String,
      active: json['active'] as bool,
      reminderEnabled: json['reminder_enabled'] as bool? ?? true,
      pillCount: json['pill_count'] as int?,
      labelImageKey: json['label_image_key'] as String?,
      labelAnalysisText: json['label_analysis_text'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  String get scheduledTimeString {
    final h = scheduledTime.hourOfPeriod == 0 ? 12 : scheduledTime.hourOfPeriod;
    final m = scheduledTime.minute.toString().padLeft(2, '0');
    final period = scheduledTime.period == DayPeriod.am ? 'AM' : 'PM';
    return '$h:$m $period';
  }

  String get timeApiFormat {
    final h = scheduledTime.hour.toString().padLeft(2, '0');
    final m = scheduledTime.minute.toString().padLeft(2, '0');
    return '$h:$m:00';
  }
}
